### DISCUSSION TITLE
Cell background colour with a gradient

### DISCUSSION THREAD

**Discussion by kiryph** [lvjr/tabularray#296] [Q&A]
I like the separation of cell style which tabularray offers:

```tex
\documentclass[border=1cm]{standalone}
\usepackage{tabularray} 
\usepackage[table]{xcolor} 
\begin{document}
\begin{tblr}{
    hlines = {white},
    vlines = {white},
    cell{1}{odd} = {teal7},
    cell{1}{even} = {green7},
    cell{2,4,6}{1} = {red7},
    cell{3,5}{1} = {purple7},
    cell{4}{3} = {azure7},
}
Alpha   & Beta  & Gamma   & Delta   \\ 
Epsilon & Zeta  & Eta     & Theta   \\
Iota    & Kappa & Lambda  & Mu      \\
Nu      & Xi    & Omicron & Pi      \\ 
Rho     & Sigma & Tau     & Upsilon \\
Phi     & Chi   & Psi     & Omega   \\
\end{tblr} 
\end{document}
```
![Screenshot 2022-08-22 at 09 50 21](https://user-images.githubusercontent.com/371756/185868175-97346bfb-8fc8-4710-9693-59e3899e285b.png)

**How can I define the background color in cell omikron, currently `cell{4}{3} = {azure7}`, to be a gradient?**

Ideally, I want to have a gradient with 45 degrees rotation. I have row and column headers which each has a different background color (A and B). The cell in the table should indicate this difference by the gradient from color A to color B. See following photoshopped screenshot:

![Screenshot 2022-08-22 at 09 50 21](https://user-images.githubusercontent.com/371756/185869408-4e92f3ec-1940-408e-842b-f99ad1b66fc7.png)

Related questions on tex.stackexchange.com

- [Gradient color in one cell of a table](https://tex.stackexchange.com/questions/59227/gradient-color-in-one-cell-of-a-table)
- [Gradient of color in one cell follow up question](https://tex.stackexchange.com/questions/59357/gradient-of-color-in-one-cell-follow-up-question)

I would prefer to stick with tabularray, since it supports multi-page tables which [nicematrix/tikz-matrix does not](https://tex.stackexchange.com/questions/618356/is-it-possible-to-use-nicematrix-with-longtable).

**Simplified Solution**

I would also be fine using two colors in one cell as following:

![Screenshot 2022-08-22 at 09 50 21](https://user-images.githubusercontent.com/371756/185871062-4d322aa9-b85c-43c1-a4d9-b3ab39bcd169.png)

**Comment by lvjr** [comment]
Not yet possible to do this, but there is already a related issue https://github.com/lvjr/tabularray/issues/159.

**Reply by kiryph** [reply]
Can you have a look at my [tikz solution](https://github.com/lvjr/tabularray/discussions/296#discussioncomment-3448121)  why `\makebox[1cm][c]` as part of `preto` and `appto` does not center text in a shaded cell?
![Screenshot 2022-08-22 at 16 43 19](https://user-images.githubusercontent.com/371756/185949598-0394f369-769d-4bb1-8e6d-71b7b0d85880.png)

**Comment by kiryph** [comment]
### Using Tikz marks and tikz shading

![Screenshot 2022-08-22 at 16 18 14](https://user-images.githubusercontent.com/371756/185943871-f5a595a9-b32f-4d18-8202-882cb7707368.png)

```tex
\documentclass{article}
\usepackage[table]{xcolor} 
\usepackage{tabularray} 
\usepackage{tikz}
\usetikzlibrary{calc}

% Andrew Stacey's code from
% http://tex.stackexchange.com/a/50054/3954
\makeatletter
\tikzset{%
  remember picture with id/.style={%
    remember picture,
    overlay,
    save picture id=#1,
  },
  save picture id/.code={%
    \edef\pgf@temp{#1}%
    \immediate\write\pgfutil@auxout{%
      \noexpand\savepointas{\pgf@temp}{\pgfpictureid}}%
  },
  if picture id/.code args={#1#2#3}{%
    \@ifundefined{save@pt@#1}{%
      \pgfkeysalso{#3}%
    }{
      \pgfkeysalso{#2}%
    }
  }
}

\def\savepointas#1#2{%
  \expandafter\gdef\csname save@pt@#1\endcsname{#2}%
}

\def\tmk@labeldef#1,#2\@nil{%
  \def\tmk@label{#1}%
  \def\tmk@def{#2}%
}

\tikzdeclarecoordinatesystem{pic}{%
  \pgfutil@in@,{#1}%
  \ifpgfutil@in@%
    \tmk@labeldef#1\@nil
  \else
    \tmk@labeldef#1,(0pt,0pt)\@nil
  \fi
  \@ifundefined{save@pt@\tmk@label}{%
    \tikz@scan@one@point\pgfutil@firstofone\tmk@def
  }{%
  \pgfsys@getposition{\csname save@pt@\tmk@label\endcsname}\save@orig@pic%
  \pgfsys@getposition{\pgfpictureid}\save@this@pic%
  \pgf@process{\pgfpointorigin\save@this@pic}%
  \pgf@xa=\pgf@x
  \pgf@ya=\pgf@y
  \pgf@process{\pgfpointorigin\save@orig@pic}%
  \advance\pgf@x by -\pgf@xa
  \advance\pgf@y by -\pgf@ya
  }%
}
\newcommand\tikzmark[2][]{%
\tikz[remember picture with id=#2] #1;}
\makeatother
% end of Andrew's code

\newcommand\ShadeCell[3]{%
  \begin{tikzpicture}[overlay,remember picture]%
    \shade[#3] ( $ (pic cs:#1) + (-6pt,10pt) $ ) rectangle ( $ (pic cs:#2) + (6pt,-6pt) $ );
  \end{tikzpicture}%
}%

\begin{document}

% all n cells with shades need a separate tikzmark startn
\ShadeCell{start1}{end1}{%
 shading = axis,left color=blue6,right color=magenta8,shading angle=135}
\ShadeCell{start2}{end2}{%
  shading = axis,left color=red7,right color=green9,shading angle=135}

\begin{tblr}{
    colspec={| l | c | c |},
    cell{1}{3}={bg=magenta8},
    cell{1}{2}={bg=green9},
    cell{2}{1}={bg=blue6},
    cell{3}{1}={bg=red7},
    cell{2}{3}={preto={\tikzmark{start1}\makebox[1cm][c]\expandafter\bgroup},appto={\expandafter\egroup\tikzmark{end1}}},
    cell{3}{2}={preto={\tikzmark{start2}\makebox[1cm][c]\expandafter\bgroup},appto={\expandafter\egroup\tikzmark{end2}}},
  }
\hline
 & A  & B \\
 \hline
C & & \\
\hline
D & & \\
\hline
\end{tblr}

\end{document}
```

based on tex.stackexchange.com

- https://tex.stackexchange.com/a/59358/8917
- https://tex.stackexchange.com/a/215917/8917

**Comment by lvjr** [comment]
If you accept the limitation that every shade cell must be the widest one in its column and the highest one in its row, then you can write your code based on the existing solution in https://github.com/CTeX-org/forum/issues/239 :

```latex
\documentclass{article}
\usepackage{tabularray}
\usepackage{tikz}

\ExplSyntaxOn

\box_new:N \l__tblr_tikz_box

\cs_new_protected:Npn \__tblr_tikzbox_fix:n #1
  {
    \hbox_set:Nn \l__tblr_tikz_box {#1}
    \box_set_ht:Nn \l__tblr_tikz_box { \box_ht:N \l__tblr_tikz_box - \abovesep }
    \box_set_dp:Nn \l__tblr_tikz_box { \box_dp:N \l__tblr_tikz_box - \belowsep }
    \box_set_wd:Nn \l__tblr_tikz_box { \box_wd:N \l__tblr_tikz_box - \rightsep }
    \skip_horizontal:n { -\leftsep }
    \box_use:N \l__tblr_tikz_box
  }

\NewContentCommand  \tikzbox [2]
  {
    \__tblr_tikzbox_fix:n
      {
        \tikz[baseline=\dim_use:N \l__tblr_strut_ht_dim + \belowsep] {#1}
        #2
      }
  }

\ExplSyntaxOff

\NewContentCommand \shadebox [3] {%
  \tikzbox{
    \useasboundingbox (0,0) rectangle (2cm,1cm);
    \pgfnodealias{cbb}{current bounding box}
    \shade[shading = axis,left color=#1,right color=#2,shading angle=135]
      (cbb.south east) rectangle (cbb.north west);
  }{%
    \llap{\makebox[2cm][c]{#3}}%
  }%
}

\begin{document}

\begin{tblr}{
  colspec={|l|c|c|}, hlines,
  cell{1}{3}={bg=magenta8},
  cell{1}{2}={bg=green9},
  cell{2}{1}={bg=blue6},
  cell{3}{1}={bg=red7},
  cell{2}{3}={cmd=\shadebox{blue6}{magenta8}},
  cell{3}{2}={cmd=\shadebox{red7}{green9}},
}
  & A & B   \\
C &   & 10c \\
D &   & 10c \\
\end{tblr}

\end{document}
```

![image](https://user-images.githubusercontent.com/4239949/186059041-bff0c1c0-2fab-4469-8564-6e4ab1768395.png)

**Comment by quark67** [comment]
There is now a solution really shorter.

And no limitation that every shade cell must be the widest one in its column and the highest one in its row.

```latex
\documentclass{article}
\usepackage{tabularray}
\UseTblrLibrary{tikz}

\begin{document}

\begin{tblrtikzbelow}
	\shade[shading = axis,left color=blue6,right color=magenta8,shading angle=135] (2-3.north west) rectangle (2-3.south east);
	\shade[shading = axis,left color=red7,right color=green9,shading angle=135] (3-2.north west) rectangle (3-2.south east);
\end{tblrtikzbelow}% no space between tblrtikzbelow and tblr
\begin{tblr}{
  colspec={|l|c|c|}, hlines,
  cell{1}{3}={bg=magenta8},
  cell{1}{2}={bg=green9},
  cell{2}{1}={bg=blue6},
  cell{3}{1}={bg=red7},
}
  & A & B   \\
C &   & 10c \\
D &   & 10c \\
\end{tblr}

\end{document}
```

<img width="203" height="156" alt="image" src="https://github.com/user-attachments/assets/17f28258-3bf2-4209-8186-26bd6583c4ce" />
